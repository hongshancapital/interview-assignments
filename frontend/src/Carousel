interface Slide {
    imageSrc: string;
    title: string;
    description: string;
  }
  
  interface SlideshowProps {
    slides: Slide[];
  }
  
  interface SlideshowState {
    currentSlideIndex: number;
  }
  
  class Slideshow extends React.Component<SlideshowProps, SlideshowState> {
    constructor(props: SlideshowProps) {
      super(props);
  
      this.state = {
        currentSlideIndex: 0,
      };
  
      this.goToNextSlide = this.goToNextSlide.bind(this);
      this.goToPreviousSlide = this.goToPreviousSlide.bind(this);
    }
  
    private goToNextSlide() {
      const { currentSlideIndex } = this.state;
      const { slides } = this.props;
  
      if (currentSlideIndex === slides.length - 1) {
        this.setState({
          currentSlideIndex: 0,
        });
      } else {
        this.setState({
          currentSlideIndex: currentSlideIndex + 1,
        });
      }
    }
  
    private goToPreviousSlide() {
      const { currentSlideIndex } = this.state;
      const { slides } = this.props;
  
      if (currentSlideIndex === 0) {
        this.setState({
          currentSlideIndex: slides.length - 1,
        });
      } else {
        this.setState({
          currentSlideIndex: currentSlideIndex - 1,
        });
      }
    }
  
    render() {
      const { slides } = this.props;
      const { currentSlideIndex } = this.state;
  
      return (
        <div className="slideshow">
          <div className="slideshow__slides">
            {slides.map((slide, index) => (
              <div
                key={index}
                className={`slideshow__slide ${
                  index === currentSlideIndex ? "slideshow__slide--active" : ""
                }`}
              >
                <img src={slide.imageSrc} alt={slide.title} />
                <div className="slideshow__slide-content">
                  <h2>{slide.title}</h2>
                  <p>{slide.description}</p>
                </div>
              </div>
            ))}
          </div>
          <button className="slideshow__previous" onClick={this.goToPreviousSlide}>
            Previous
          </button>
          <button className="slideshow__next" onClick={this.goToNextSlide}>
            Next
          </button>
        </div>
      );
    }
  }
  
  const slides: Slide[] = [
    {
      imageSrc: "https://picsum.photos/id/100/800/400",
      title: "Slide 1",
      description: "This is the first slide",
    },
    {
      imageSrc: "https://picsum.photos/id/200/800/400",
      title: "Slide 2",
      description: "This is the second slide",
    },
    {
      imageSrc: "https://picsum.photos/id/300/800/400",
      title: "Slide 3",
      description: "This is the third slide",
    },
  ];
  
  ReactDOM.render(<Slideshow slides={slides} />, document.getElementById("root"));
  
