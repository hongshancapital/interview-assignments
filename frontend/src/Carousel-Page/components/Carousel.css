

.swiper {
    width: 100%;
    background-color: blueviolet;
    overflow: hidden;
    position: relative;
    cursor: pointer;
    
  }
  
  
  
  
  
  
  
  
  
  
  
  
  .swiper .swipers {
    display: flex;
    text-align: center;
    color: #fff;
    font-size: 20px;
    
  }
  .swiper .swipers .swipers-item {
    width: 100%;
    
  }
  .swiper .swipers .swipers-item img {
    width: 100%;
  }
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  .pagnation {
    position: absolute;
    right: 0;
    left: 0;
    z-index: 15;
    display: flex !important;
    justify-content: center;
    margin-right: 15%;
    margin-left: 15%;
    padding-left: 0;
    list-style: none;
    bottom: 12px;
    
  }
  .pagnation div {
    position: relative;
    display: inline-block;
    flex: 0 1 auto;
    box-sizing: content-box;
    width: 36px;
    height: 3px;
    margin: 0 3px;
    padding: 0;
    text-align: center;
    text-indent: -999px;
    vertical-align: top;
    transition: all 0.5s;
    background: #ccc;
  }
  .animation .active span {
    display: block;
    content: " ";
    height: 100%;
    background: rgb(102, 101, 101);
  }
  
  @keyframes myfirst {
    0% {
      width: 0;
    }
    100% {
      width: 100%;
    }
  }
  