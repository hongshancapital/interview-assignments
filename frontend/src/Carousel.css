.carousel-container {
    height: 100vh;
    position: relative;
    overflow: hidden;
}

.carousel-item-wrapper {
    width: 100%;
    height: 100%;
    position: absolute;
    transition: 0.5s transform ease-in-out;
}

.indicator-box {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 30px;
    position: fixed;
    bottom: 20px;
    width: 100%;
    text-align: center;
}

.indicator-item {
    width: 50px;
    height: 3px;
    background-color: #ccc;
    cursor: pointer;
}

.indicator-item:hover {
    height: 5px;
}

.indicator-item:not(:nth-of-type(1)) {
    margin-left: 10px;
}

.indicator-item-progress {
    height: 100%;
    background: white;
    animation: progress 2s infinite;
}

@keyframes progress {
    from { width: 0; }
    to { width: 100%; }
}

