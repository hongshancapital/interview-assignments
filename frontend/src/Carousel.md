### Carousel API

---
#### Props
| 参数                 | 类型                                              | 默认值  | 必填  | 说明               |
|--------------------|-------------------------------------------------|------|-----|------------------|
| slides             | Slide[]                                         | -    | 是   | 幻灯片对象集合          |
| duration           | number                                          | 3000 | 否   | 幻灯片自动切换超时时间      |
| activeId           | number                                          | -    | 否   | 当前显示的幻灯片ID（受控模式） |
| onClickSlideMarker | (id: string &#124; number, idx: number) => void | -    | 否   | 当点击指示器时的回调       |
---
#### Type: Slide
| 属性              | 类型                    | 是否必填 | 描述          |
|-----------------|------------------------|------|-------------|
| id              | string &#124; number   | 否    | 用于受控模式下的唯一标识 |
| textInfos       | TextInfo[]             | 否    | 文本信息集合      |
| backgroundImage | string                 | 是    | 背景图片        |
| extStyles       | Record<string, string> | 否    | 拓展样式 |
---
### Type: TextInfo
| 属性     | 类型                                      | 是否必填 | 默认值     | 描述               |
|--------|-----------------------------------------|------|---------|------------------|
| type   | 'title' &#124;  'desc' &#124;  'custom' | 否    | 'title' | 文本类型 - 标题、正文、自定义 |
| text   | string                                  | 否    | -       | 文本内容，使用自定义渲染时可不传 |
| render | () => React.ReactNode                   | 否    | -       | 自定义渲染函数          |
