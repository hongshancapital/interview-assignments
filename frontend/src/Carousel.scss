.item-wrapper {
  white-space: nowrap;
  transition: transform 0.3s;
}

.carousel-item {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  height: 200px;
  color: #fff;
  background-color: #000;
}

.progress {
  position: absolute;
  bottom: 0;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  margin-bottom: 10px;
  width: 100%;
  .bar-wrapper {
    width: 90px;
    height: 7px;
    background-color: gray;
    margin-left: 20px;
    cursor: pointer;
    .bar {
      animation-name: progressBar;
      height: 100%;
      animation-fill-mode: forwards;
      animation-iteration-count: infinite;
    }
  }
  @keyframes progressBar {
    0% {
      width: 0%;
    }
    100% {
      width: 100%;
    }
  }
}
.card {
  display: flex;
  flex-direction: column;
  align-items: center;
  .img-wrapper {
    width: 150px;
    height: 150px;
    img {
      object-fit: contain;
      width: 100%;
      height: 100%;
    }
  }
}
