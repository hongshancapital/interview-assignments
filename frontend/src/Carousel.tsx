import {useState, useEffect} from 'react';
import './App.css';
import {Progress} from 'antd';
import 'antd/dist/antd.css';
import XPhone from './componentsFor/XPhone';
import Tablet from './componentsFor/Tablet';
import Buy from './componentsFor/Buy';


function App() {
   const [percentage, setPercentage] = useState(0);
   const [fstPercentage, setFstPercentage] = useState(0);
   const [sedPercentage, setSedPercentage] = useState(0);
   const [trdPercentage, setTrdPercentage] = useState(0);

  useEffect(() => {
    
     setTimeout(() => {
      let p = percentage;
      ++p;
      if(p > 300) {
        p = p%300;
      }
      setPercentage(p);
      if(percentage <= 300) {
        if(percentage <= 100) {
          let fstPercentage = percentage;
          setFstPercentage(fstPercentage);
          setSedPercentage(0);
          setTrdPercentage(0);
        }
        else if(percentage > 100 && percentage <= 200) {
          let scdPercentage = percentage - 100;
          setFstPercentage(100);
          setSedPercentage(scdPercentage);
          setTrdPercentage(0);
        }
        else if(percentage > 200 && percentage <= 300) {
          let trdPercentage = percentage - 200;
          setFstPercentage(100);
          setSedPercentage(100);
          setTrdPercentage(trdPercentage);
        }
      }
     }, 50);
  }, [percentage])

  return (
    <div className = {(percentage >=0 && percentage <= 100) ? 'background__black' : 'background__normal'}>
          <div className="main__content">
              {(percentage >=0 && percentage <= 100) && <XPhone/>}
              {(percentage <= 200 && percentage >= 100) && <Tablet/>}
              {(percentage <= 300 && percentage > 200) && <Buy/>}
          </div>

        <div id="container"> 
          <Progress   percent={percentage} showInfo={false}/> 

          <Progress    percent={sedPercentage} showInfo={false}/>

          <Progress    percent={trdPercentage} showInfo={false}/>
        </div>
    </div>
  );
}

export default App;
