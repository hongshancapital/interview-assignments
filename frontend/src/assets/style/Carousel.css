.carousel {
  position: relative;
  height: 100vh;
  overflow: hidden;
}

.carousel__wrapper {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  transition: all .5s ease-in-out;
}

.wrapper__page {
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  background-color: #f1f1f3;
}

.wrapper__page:first-child {
  background-color: #090909;
}

.wrapper__page:nth-child(2) {
  background-color: #fafafa;
}

.page__title > .title {
  font-size: 60px;
  margin: 0;
  white-space: pre-line;
}

.page__title > .subtitle {
  font-size: 30px;
  margin-top: 10px;
  white-space: pre-line;
}

.wrapper__page:first-child > .page__title {
  color: #fff;
}

.wrapper__page > img {
  width: 150px;
  height: 150px;
}

.wrapper__page > p {
  font-size: 40px;
  font-weight: bold;
}

.carsousel__indiactor {
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
  bottom: 100px;
  display: flex;
  align-items: center;
  justify-content: center;
}

.indiactor__bar {
  width: 40px;
  height: 3px;
  border-radius: 1.5px;
  background-color: rgba(169, 169, 169, 1);
  position: relative;
  cursor: pointer;
}

.indiactor__bar:hover {
  transform: scaleY(2)
}

.indiactor__bar + .indiactor__bar {
  margin-left: 10px;
  overflow: hidden;
  border-width: 0;
}

.bar__fill {
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  background-color: #fff;
  border-radius: 2px;
  width: 0;
}
