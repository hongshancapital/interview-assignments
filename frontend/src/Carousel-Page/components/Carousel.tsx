import React, { useRef, useEffect, FC, useState, useCallback } from "react";
import "./Carousel.css";

interface Types {
  children: React.ReactNode;
  resourceList: Array<any>;
  direction?: "left" | "right";
  timeout?: number;
}



const Carousel: FC<Types> = ({ children, direction = "left", timeout = 3000, resourceList }) => {

  let timer: any = null;
  let pageWidth: number = 0;
  let disX: number = 0;
  let X: number = 0;
  let page: number = 0;
  let startX: number = 0;
  let endX: number = 0;



  {
    React.Children.map(children, (element: any, idx) => {
      let el = React.cloneElement(element, { ref: idx });
    })
  }
  const dragBox = useRef<HTMLDivElement | null>(null);

  const swiper = useRef<HTMLDivElement | null>(null);

  const pagnation = useRef<HTMLDivElement | null>(null);



  const getChildren = () => children as any;

  const getChildrenLength = getChildren().length;

  const getdragPagnation = () => pagnation.current as any;

  const getdragSwiper = () => swiper.current as any;

  const getdragBox = () => dragBox.current as any;

  const [pageIndex, setPageIndex] = useState<number>(0);
  const setDrag = useCallback((X: number) => {
    getdragBox().style.transform = `translate3d(${X - pageWidth}px,0px,0px)`;
  }, []);

  const fnE = useCallback(() => {
    if (page <= -1) {
      page = 4;
      setPageIndex(page);
      getdragBox().style.transition = "none";
      X = -page * pageWidth;
      setDrag(X);
    }
    if (page >= getChildrenLength) {
      page = 0;
      setPageIndex(page);
      getdragBox().style.transition = "none";
      X = -page * pageWidth;
      setDrag(X);
    }
  }, [setDrag, getChildrenLength]);

  const leftDrag = useCallback(
    (index?: number) => {
      getdragBox().style.transition = ".3s ease all";
      page = index ?? page + 1;
      setPageIndex(page);
      X = -page * pageWidth;
      getdragBox().ontransitionend = fnE;
      setDrag(X);
    },
    [fnE, setDrag]
  );
  const rightDrag = useCallback(
    (index?: number) => {
      getdragBox().style.transition = ".3s ease all";
      page = index ?? page - 1;
      setPageIndex(page);
      X = -page * pageWidth;
      getdragBox().ontransitionend = fnE;
      setDrag(X);
    },
    [fnE, setDrag]
  );
  const FnEnd = useCallback(
    (e: MouseEvent) => {
      endX = e.pageX;
      if (Math.abs(startX - endX) > 80) {
        startX > endX ? leftDrag() : rightDrag();
      } else {
        getdragBox().style.transition = ".3s ease all";
        X = -page * pageWidth;
        getdragBox().ontransitionend = fnE;
        setDrag(X);
      }
      getdragBox().onmousemove = null;
      getdragBox().onmouseup = null;
    },
    [fnE, leftDrag, rightDrag, setDrag]
  );

  const FnMove = useCallback(
    (e: MouseEvent) => {
      X = e.pageX - disX;
      setDrag(X);
    },
    [setDrag]
  );

  const fnMove = useCallback(() => {
    timer = setInterval(() => {
      direction === "left" ? leftDrag() : rightDrag();
    }, timeout);
    getdragPagnation().className = "pagnation animation";
  }, [direction, leftDrag, rightDrag, timeout]);

  const FnStart = useCallback(
    (e: MouseEvent) => {
      getdragBox().style.transition = "none";
      startX = e.pageX;
      disX = startX - X;
      getdragBox().onmousemove = FnMove;
      getdragBox().onmouseup = FnEnd;
      return false;
    },
    [FnEnd, FnMove]
  );

  const fnOver = useCallback(() => {
    clearInterval(timer);
    getdragPagnation().className = "pagnation";
  }, []);

  const fnTo = (index: number) => {
    const cur = index;
    pageWidth = getdragSwiper().clientWidth;
    clearInterval(timer);
    direction === "left" ? leftDrag(cur) : rightDrag(cur);
    // fnMove();
  };

  useEffect(() => {
    pageWidth = getdragSwiper().clientWidth;
    getdragBox().style.width = `${pageWidth * (getChildrenLength + 2)}px`;
    getdragBox().style.transform = `translate3d(${-pageWidth}px,0px, 0px)`;
    getdragBox().onmousedown = FnStart;
    getdragBox().onmouseover = fnOver;
    getdragPagnation().onmouseover = fnOver;
    getdragBox().onmouseout = fnMove;
  }, [FnStart, fnOver, fnMove, getChildrenLength]);

  useEffect(() => {
    fnOver();
    fnMove();
  }, [fnOver, fnMove]);



  return (
    <div style={{ position: 'relative' }}>
      <div className="swiper" ref={swiper}>
        <div className="swipers" ref={dragBox}>
          <div className="swipers-item" style={resourceList[getChildrenLength - 1].style}>
            {getChildren()[getChildrenLength - 1]}
          </div>
          {getChildren().map((item: any, index: any) => (
            <div key={index} className="swipers-item" style={resourceList[index].style}>
              {getChildren()[index]}
            </div>
          ))}
          <div className="swipers-item" style={resourceList[0].style}>
            {getChildren()[0]}
          </div>
        </div>
      </div>
      <div className="pagnation" ref={pagnation}>
        {getChildren().map((item: any, index: any) => (
          <div onClick={() => fnTo(index)} key={index} className={pageIndex === index ? "active" : ""}>
            <span style={{ animation: pageIndex === index ? `myfirst ${timeout / 1000}s linear` : 'initial' }}></span>
          </div>
        ))}
      </div>
    </div>
  );
};

export default Carousel;
