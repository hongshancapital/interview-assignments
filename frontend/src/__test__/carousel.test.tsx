import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { CSSProperties, createRef } from 'react'

import Carousel, { CarouselRef } from '../components/Carousel'

const imgs = [
  'https://images.unsplash.com/photo-1562886877-aaaa5c0b3225?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjk5Mjc0fQ',
  'https://images.unsplash.com/photo-1572652963245-bd7fda887078?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjk5Mjc0fQ',
  'https://images.unsplash.com/photo-1572656934803-d2162b2e98bf?ixlib=rb-1.2.1&q=80&fm=jpg&crop=entropy&cs=tinysrgb&w=1080&fit=max&ixid=eyJhcHBfaWQiOjk5Mjc0fQ'
]

const divStyle: CSSProperties = {
  height: '400px',
  background: '#364d79'
}

const imgStyle: CSSProperties = {
  width: '100%',
  height: '100%'
}

describe('Carousel Test', () => {
  const ref = createRef<CarouselRef>()
  beforeEach(() => {
    render(
      <Carousel ref={ref}>
        {imgs.map((img: string, index: number) => (
          <div style={divStyle} key={index}>
            <img src={imgs[index]} style={imgStyle} alt={img} data-testid='img' />
          </div>
        ))}
      </Carousel>
    )
  })

  it('Carousel Visible test', async () => {
    expect(await screen.findByTestId('carousel')).toBeVisible()
  })

  it('Slide Count test', async () => {
    const slideContentCount = imgs.length
    const result = await screen.findAllByTestId('img')
    expect(result).toHaveLength(slideContentCount)
  })

  it('Click test', async () => {
    const allPaginationItems = await screen.findAllByTestId('pagination-item')
    fireEvent.click(allPaginationItems[0])
    await waitFor(async () => {
      expect(await screen.findByAltText(imgs[2])).toBeVisible()
    })
  })

  it('Should have prev,next and go function', async () => {
    const { prev, next, goTo } = ref.current as CarouselRef
    expect(typeof prev).toBe('function')
    expect(typeof next).toBe('function')
    expect(typeof goTo).toBe('function')
  })
})
