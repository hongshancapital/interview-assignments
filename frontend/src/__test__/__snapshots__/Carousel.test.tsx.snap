// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel test Carousel rendered 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="innerWrapper"
      data-testid="carousel-inner-wrapper"
      style="transform: translateX(-0%);"
    />
    <div
      class="wrapper"
      data-testid="indicator-wrapper"
    />
  </div>
</div>
`;

exports[`Carousel test carousel rendered 1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="innerWrapper"
      data-testid="carousel-inner-wrapper"
      style="transform: translateX(-0%);"
    />
    <div
      class="wrapper"
      data-testid="indicator-wrapper"
    />
  </div>
</div>
`;

exports[`Carousel test click Indicator of Carousel  1`] = `
<div>
  <div
    class="wrapper"
  >
    <div
      class="innerWrapper"
      data-testid="carousel-inner-wrapper"
      style="transform: translateX(-100%);"
    >
      <div
        class="carouselItem"
      >
        text1
      </div>
      <div
        class="carouselItem"
      >
        text2
      </div>
    </div>
    <div
      class="wrapper"
      data-testid="indicator-wrapper"
    >
      <div
        class="item"
      >
        <div
          class=""
          style=""
        />
      </div>
      <div
        class="item"
      >
        <div
          class="activeItem"
          style="transition-duration: 1000ms;"
        />
      </div>
    </div>
  </div>
</div>
`;
